### PR TITLE
allow rhs in rref

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

#25661 related

#### Other comments

see issue page and the questions I raise there

#### Release Notes

* matrices
  * `A.rref_rhs(b)` will return rref of A and resulting form of b without reducing b; pivots are returned as 3rd arg

### CODE REVIEW

File: sympy/matrices/tests/test_reductions.py [hunk 1]
```diff
@@ -279,6 +280,20 @@ def test_rref():
...
+    rhs = Matrix([a, b, c, d])
+    assert A.rref(rhs=rhs) == (Matrix([
+    [1, 0],
+    [0, 1],
+    [0, 0],
+    [0, 0]]), Matrix([
+    [   -2*c + d],
+    [3*c/2 - d/2],
+    [          a],
+    [          b]]), (0, 1, 2, 3))
```
Comment (sylee957) [inline sympy/matrices/tests/test_reductions.py:294]: The indices of the pivots should be (0, 1)

Comment (smichr) [reply sympy/matrices/tests/test_reductions.py]: I returned rows instead of cols. Fixed.

File: sympy/matrices/reductions.py [hunk 2]
```diff
@@ -356,13 +366,20 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
     [0, 1, -0.712328767123288, 0],
     [0, 0,         0,          0]]), (0, 1))
 
+    >>> Matrix([[1, 1], [2, 1]]).rref(rhs=Matrix(symbols('r1 r2')))
+    (Matrix([
+    [1, 0],
+    [0, 1]]), (0, 1), Matrix([
+    [ -r1 + r2],
+    [2*r1 - r2]]))
```
Comment (sylee957) [inline sympy/matrices/reductions.py:374]: The doctest is not correct

File: sympy/matrices/reductions.py [hunk 3]
```diff
@@ -356,13 +366,20 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
     [0, 1, -0.712328767123288, 0],
     [0, 0,         0,          0]]), (0, 1))
 
+    >>> Matrix([[1, 1], [2, 1]]).rref(rhs=Matrix(symbols('r1 r2')))
```
Comment (sylee957) [inline sympy/matrices/reductions.py:369]: I also suggest to create the matrix `Matrix([r1, r2])`.

**Comment by oscarbenjamin** [comment]
What about having a separate function like `rref_aug` or `rref_rhs` or something?

**Comment by sylee957** [comment]
I wanted to vote about the idea to have separate function, especially if the structure of return is dependent on some parameters.

**Comment by donnaaboise** [comment]
Thanks for SymPy developers for looking into this! 

I haven't actually tested the PR, but do have the following questions/comments : 

* Does this PR handle the case that @smichr pointed out?   That is, the case in the pivot rows are not the first `r=A.rank()` rows?   A fix would be to allow row exchanges.  It seems that A.rref() already does row exchanges and can handle, for example, Matrix([[0,0],[0,1]]).   Can this PR get at the row permutations needed?  (my original approach failed with this matrix).

File: sympy/matrices/matrices.py [hunk 4]
```diff
@@ -176,6 +176,30 @@ def is_echelon(self):
     def rank(self, iszerofunc=_iszero, simplify=False):
         return _rank(self, iszerofunc=iszerofunc, simplify=simplify)
 
+    def rref_rhs(self, rhs, iszerofunc=_iszero, simplify=False, pivots=True,
+        normalize_last=True):
```
Comment (oscarbenjamin) [inline sympy/matrices/matrices.py:180]: In new methods I would prefer not to add all of these options. I regret that they are already there in the existing functions but we should not add them in new functions unless there is some significant reason.

Here `_iszero` should be handled by `DomainMatrix`. It is a bad idea to use `simplify`. The `normalize_last` parameter will not make any sense of `DomainMatrix` is being used. The `pivots` return value was always a mistake at least it should never have been included in end user API.

I would reduce this signature to:
```python
def rref_rhs(self, rhs):
```
and not return the pivots.

**Comment by smichr** [comment]
> Thanks for SymPy developers for looking into this!
> 
> I haven't actually tested the PR, but do have the following questions/comments :
> 
> * Does this PR handle the case that @smichr pointed out?   That is, the case in the pivot rows are not the first `r=A.rank()` rows?   A fix would be to allow row exchanges.  It seems that A.rref() already does row exchanges and can handle, for example, Matrix([[0,0],[0,1]]).   Can this PR get at the row permutations needed?  (my original approach failed with this matrix).

Yes, this case is handled and the rank calculation is omitted. @sylee957 pointed out that simply satcking a full identity matrix is a more robust aproach.

File: sympy/matrices/matrices.py (hunk 4 above)
Comment (smichr) [reply sympy/matrices/matrices.py]: done

File: sympy/matrices/reductions.py [hunk 5]
```diff
@@ -320,6 +321,14 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
         each pivot is normalized to be `1` before row operations are
         used to zero above and below the pivot.
 
+    rhs : Matrix
+        If ``rhs`` is passed it must have the same number of rows as ``M``.
+        Reduction steps will not take place in this matrix, but
+        the elements will be included in the reduction steps. If a single
+        column matrix of symbols is passed, the expressions in each
+        row at the end of the reduction will indicate how rows were
+        combined to perform the reduction.
```
Comment (sylee957) [inline sympy/matrices/reductions.py:330]: Unfortunately, I missed this point that this should move to the documentation in `rref_rhs`

Comment (smichr) [reply sympy/matrices/reductions.py:330]: I will make new pr. Sorry I missed that.

**Comment by oscarbenjamin** [comment]
Thanks everyone!

**Pull request merged**